XenAdmin/Program.cs

```diff
@@ -990,9 +998,59 @@ public static void ReconfigureConnectionSettings()
                 Properties.Settings.Default.ProvideProxyAuthentication = false;
                 Properties.Settings.Default.Save();
             }
+            else
+            {
+                ReconfigureProxyAuthenticationSettings();
+            }
             XenAPI.Session.Proxy = XenAdminConfigManager.Provider.GetProxyFromSettings(null);
         }
 
+        /// <summary>
+        /// Stores the Basic and Digest authentication modules, used for proxy server authentication, 
+        /// for later use; this is needed because we cannot create new instances of them and it 
+        /// saves us needing to create our own custom authentication modules.
+        /// </summary>
+        private static void RememberProxyAuthenticationModules()
+        {
+            var authModules = AuthenticationManager.RegisteredModules;
+            while (authModules.MoveNext())
+            {
+                var module = (IAuthenticationModule)authModules.Current;
+                if (module.AuthenticationType == "Basic")
+                    BasicAuthenticationModule = module;
+                else if (module.AuthenticationType == "Digest")
+                    DigestAuthenticationModule = module;
+            }
+        }
+
+        /// <summary>
+        /// Configures .NET's AuthenticationManager to only use the authentication module that is 
+        /// specified in the ProxyAuthenticationMethod setting. Also sets XenAPI's HTTP class to 
+        /// use the same authentication method.
+        /// </summary>
+        private static void ReconfigureProxyAuthenticationSettings()
+        {
+            var authModules = AuthenticationManager.RegisteredModules;
+            var modulesToUnregister = new List<IAuthenticationModule>();
+
+            while (authModules.MoveNext())
+            {
+                var module = (IAuthenticationModule)authModules.Current;
+                modulesToUnregister.Add(module);
+            }
+
+            foreach (var module in modulesToUnregister)
+                AuthenticationManager.Unregister(module);
+
+            var authSetting = (HTTPHelper.ProxyAuthenticationMethod)Properties.Settings.Default.ProxyAuthenticationMethod;

```

Comment:
What will happen here if the ProxyAuthenticationMethod value is different than any value in the HTTPHelper.ProxyAuthenticationMethod enum (for example if someone changes it  manually to 2)?